storage/innobase/dict/dict0dict.cc

```diff
@@ -7450,27 +7450,93 @@ dberr_t dict_get_dictionary_info_by_id(ulint dict_id, char **name,
   return err;
 }
 
-bool dict_detect_encryption(bool is_upgrade, space_id_t mysql_plugin_space) {
-  bool encrypt_mysql = false;
-  if (is_upgrade) {
-    if (mysql_plugin_space == SYSTEM_TABLE_SPACE) {
-      return (srv_sys_space.is_encrypted());
-    }
+/** Detect if database has encrypted mysql plugin space. This is
+ *  indication that system tablespace was encrypted.
+@param[in] mysql_plugin_space   space_id of mysql/plugin table.
+@return true if encrypted, false if not (or fail to open mysql/plugin table) */
+static bool dict_is_mysql_plugin_space_encrypted(
+    space_id_t mysql_plugin_space) {
+  if (mysql_plugin_space == SYSTEM_TABLE_SPACE) {
+    return (srv_sys_space.is_encrypted());
+  }
+
+  space_id_t space_id = fil_space_get_id_by_name("mysql/plugin");
+  if (space_id == SPACE_UNKNOWN) {
+    return (false);
+  }
 
-    space_id_t space_id = fil_space_get_id_by_name("mysql/plugin");
-    if (space_id == SPACE_UNKNOWN) {
-      return (false);
-    }
+  const fil_space_t *space = fil_space_get(space_id);
+  ut_ad(space != nullptr);
+
+  if (space == nullptr) {
+    return false;
+  }
+  return FSP_FLAGS_GET_ENCRYPTION(space->flags);
+}
+
+/** Detect if innodb_encrypt_tables is set either to ON or FORCE
+@return true if innodb_encrypt_tables is equal to ON or FORCE */
+static bool dict_is_encrypt_tables_either_on_or_force() {
+  return srv_encrypt_tables == SRV_ENCRYPT_TABLES_ON ||
+         srv_encrypt_tables == SRV_ENCRYPT_TABLES_FORCE;
+}
 
-    fil_space_t *space = fil_space_get(space_id);
-    ut_ad(space != nullptr);
+/** Detect if mysql.ibd's page0 has encryption flag set.
+ *  The page 0 is read from disk.
+@param[in] filepath path of mysql.ibd
+@param[out] success true if no error
+@return true if encryption flag is set, false otherwise */
+static bool dict_mysql_ibd_page_0_has_encryption_flag_set(const char *filepath,
+                                                          bool &success) {
+  pfs_os_file_t file = os_file_create_simple_no_error_handling(
```

Comment:
You should avoid direct reads if possible. (may be buf_page_get_gen()?)

Is the double write buffer recovery over at this stage? Does the double write buffer do "sync" io on the tablespace? -> If the answer is yes to both of them, then you may use direct read.

Reply:
I have implemented reading from the buffer, as you said, but it is not possible to do so at this stage - because the fil_system is not yet ready. When I tried to acquire the tablespace I got missing tablespace error and when I tried to read page from a buffer without acquiring space I got a core dump - as assert that page's space must exists in the fil_system was false.

Reply:
I have done lot of debugging and code study and I now am in better position :)

In my debugging, I found that recovery (recv_recovery_from_checkpoint_finish()) is over before dict_detect_encryption_of_mysql_ibd().

So there shouldn't be any issues.

1. Why is fil_system not ready. If redo recovery is over, then fil_system is fully initialized.
2. Why is mysql.ibd tablespace missing in fil_system cache?
3. buf_page_get_gen() should work fine at this stage. Infact recovery also uses buffer pool read to apply redo on pages. See buf_read_recv_pages().

And to answer the questions I asked.
"Is the double write buffer recovery over at this stage? 
Yes, the dblwr buffer recovery is over before even redo starts applying/scanning.

Does the double write buffer do "sync" io on the tablespace?
Yes, The IO is sync IO and also fsync() is done.

Neverthless, based on my debugging and observations, it is not required to do direct IO. A simple buffer pool read should work.

Reply:
Ok, after our discussion our on slack and then verifying with the branch, there is *indeed* a case where we cannot do buffer pool read. This is the non-recovery case. But there is still one issue.

If recovery fixed mysq.ibd, all the changes to page 0 of mysql are done via buffer pool and regular IO.

If you do direct IO in this case, you may read a stale copy of page. The updated page 0 copy is in buffer pool (and IO threads may write at leisure).

Reply:
@satya-bodapati I will then implement first reading from buffer pool and then if the page does not exist I will read with regular IO

Reply:
Yes, May be you should use  is_mysql_ibd_page_0_in_redo() (save state). If this is true, then fil_space_t* should be present and you should do buffer pool read, else { do direct IO read)